Replacing copy-pasted ZXing with subtree

YohDeadfall opened the pull request:
Follows the spirit of #10463, but without submodules.

---

# About subtree

Read [`git subtree` article](https://www.atlassian.com/git/tutorials/git-subtree) for all details or the following text for an excerpt from the artice.

## What is `git subtree`, and why should I use it?

`git subtree` lets you nest one repository inside another as a sub-directory. It is one of several ways Git projects can manage project dependencies.

![image](https://user-images.githubusercontent.com/10694013/231784919-c7476155-0fa6-468a-9aea-19d77320d376.png)

Why you may want to consider `git subtree`

- Management of a simple workflow is easy.
- Older version of Git are supported (even older than v1.5.2).
- The sub-project’s code is available right after the clone of the super project is done.
- `git subtree` does not require users of your repository to learn anything new. They can ignore the fact that you are using `git subtree` to manage dependencies.
- `git subtree` does not add new metadata files like git submodule does (i.e., .gitmodule).
- Contents of the module can be modified without having a separate repository copy of the dependency somewhere else.

Drawbacks (but in our opinion they're largely acceptable):

- You must learn about a new merge strategy (i.e. `git subtree`).
- Contributing code back upstream for the sub-projects is slightly more complicated.
- The responsibility of not mixing super and sub-project code in commits lies with you.

# Summary

The overview shows that `git subtree` is close to the copy/paste strategy which is used at the moment by the project, but it also improves some part of the workflow like recording the original commit ID.

`git subtree` requires less actions than having a fork with NuGet packages which is a great maintainability advantage too. Only one repository is needed and no packages at all.

`git subtree` is much friendlier for any contributor than `git submodule`.

Therefore, `git subtree` sounds like the most well balanced solution being an automated tool for updating security critical references.

# Overview

The overview contains a summary of all follow up sections. It's split into two parts, advantages and disadvantages for easy understanding.

## Advantages

A advantage is highlighted by **bold** if it's present in the corresponding method.

|  | Copy/Paste | Subtree | Fork and Submodule | Fork and NuGet |
|:-|:----------:|:-------:|:------------------:|:--------------:|
| Preserves the original commit ID                   | No      | **Yes** | **Yes** | **Yes** |
| Preserves the original code the current repository | **Yes** | **Yes** | No      | No      |
| Allows to fork and build the current repository    | **Yes** | **Yes** | No      | **Yes** |
| Allows modifications to the original code          | **Yes** | **Yes** | **Yes** | **Yes** |
| Actions to add a reference (more is worse)         | 4       | **3**   | **3**   | 6       |
| Actions to update a reference (more is worse)      | 4       | **3**   | 6       | 7       |

## Disadvantages

A disadvantage is highlighted by **bold** if it's present in the corresponding method.

|  | Copy/Paste | Subtree | Fork and Submodule | Fork and NuGet |
|:-|:----------:|:-------:|:------------------:|:--------------:|
| Requires modification to the original code to use  | No      | No      | No      | **Yes** |
| Affects local build time                           | **Yes** | **Yes** | **Yes** | No      |
| Affects CI build time                              | **Yes** | **Yes** | **Yes** | No      |
| Actions to add a reference (more is worse)         | 4       | 3       | 3       | **6**   |
| Actions to update a reference (more is worse)      | 4       | 3       | **6**   | **7**   |

# Detailed comparison

## Preserves the original commit ID

| Method | Description |
|:-------|:------------|
| Copy/Paste         | **No**, if not added somewhere manually |
| Subtree            | **Yes**, an automatically generated commit message contains it |
| Fork and Submodule | **Yes**, commits match the original repo or contained in merge commits |
| Fork and NuGet     | **Yes**, commits match the original repo or contained in merge commits |

## Preserves the original code the current repository

| Method | Description |
|:-------|:------------|
| Copy/Paste         | **Yes**, code is incorporated into the current repository |
| Subtree            | **Yes**, code is incorporated into the current repository |
| Fork and Submodule | **No**, code must kept in a separate repository |
| Fork and NuGet     | **No**, code must kept in a separate repository |

## Allows to fork and build the current repository

| Method | Description |
|:-------|:------------|
| Copy/Paste         | **Yes**, the repository can be forked/cloned and built immediately |
| Subtree            | **Yes**, the repository can be forked/cloned and built immediately |
| Fork and Submodule | **No**, the repository requires submodules to be initialized first  |
| Fork and NuGet     | **Yes**, the repository can be forked/cloned and built immediately |

## Allows modifications to the original code

"Original code" means code stored in the original repository of a project being used by Wasabi.

| Method | Description |
|:-------|:------------|
| Copy/Paste         | **Yes**, code can be modified without an ability to contibute back |
| Subtree            | **Yes**, code can be modified with an ability to contribute back, but it's complicated |
| Fork and Submodule | **Yes**, code can be modified with an ability to contribute back |
| Fork and NuGet     | **Yes**, code can be modified with an ability to contribute back|

## Requires modification to the original code to use

| Method | Description |
|:-------|:------------|
| Copy/Paste         | **No**, code can be used as is if it compiles |
| Subtree            | **No**, code can be used as is if it compiles and the referenced project has the required TFM |
| Fork and Submodule | **No**, code can be used as is if it compiles and the referenced project has the required TFM ||
| Fork and NuGet     | **Yes**, projects in a forked repository must be changed to include `WalletWasaby` prefix a package name prefix to avoid potential legal issues with the original author after publishing them to a public NuGet source like NuGet.org |

## Affects local build time

Copy-pasted code tends to be as minimal as possible to solve only a single problem or a few. Subtree and forks allow to reduce code size too, but according to the rule of thumb forks usually contain unmodified code or code with custom patches.

| Method | Description |
|:-------|:------------|
| Copy/Paste         | **Yes**, code is incorporated and must be compiled, but time depends on amount of the original code being preserved |
| Subtree            | **Yes**, code is incorporated and must be compiled, but time depends on amount of the original code being preserved |
| Fork and Submodule | **Yes**, code is referenced and must be compiled, but time depends on amoout of the original code being preserved |
| Fork and NuGet     | **No**, code is compiled once and then can be resolved from a NuGet source |

## Affects CI build time

The same comment as in the previous section applies here, i.e. less code - less time it takes to compile it.

| Method | Description |
|:-------|:------------|
| Copy/Paste         | **Yes**, code is incorporated and must be compiled, but time depends on amount of the original code being preserved |
| Subtree            | **Yes**, code is incorporated and must be compiled, but time depends on amount of the original code being preserved |
| Fork and Submodule | **Yes**, code is referenced and must resolved through `submodule init` and then be compiled, but time of the second step depends on amoout of the original code being preserved |
| Fork and NuGet     | **No**, code is compiled once and then can be resolved from a NuGet source |

## Actions to add a reference

### Copy/Paste

1. Clone or download the original repository.
2. Copy required code into WalletWasabi repository.
3. Reference it as a project or include files into an existing Wasabi project.
4. Write the original commit ID.

### Subtree
1. Execute `git subtree add`.
2. Remove non-required code if needed.
3. Reference it as a project or include files into an existing Wasabi project.

### Fork and Submodule

1. Make a fork into zkSNACKs organization (admin rights required).
2. Remove non-required code if needed.
3. Execute `git submodule add`.

### Fork and NuGet

1. Make a fork into zkSNACKs organization (admin rights required).
2. Remove non-required code if needed.
3. Add `WalletWasabi` prefix to projects.
4. Build code.
5. Publish packages.
6. Add packages in the required WasabiWallet project.

## Actions to update a reference

### Copy/Paste
1. Clone or download the original repository.
2. Copy required code into WalletWasabi repository.
3. Write the original commit ID.
4. Open a pull request for changes to be reviewed.

###Subtree

1. Execute `git subtree update`.
2. Resolve conflicts if they are.
3. Open a pull request for changes to be reviewed.

### Fork and Submodule

1. Fetch the original repository.
2. Resolve conflicts if the are.
3. Open a pull request for changes to be reviewed.
5. Execute `git submodule update` in WalletWasabi repository.
6. Notfy teammates about the update (optional).

### Fork and NuGet

1. Fetch the original repository.
2. Resolve conflicts if the are.
3. Open a pull request for changes to be reviewed.
5. Build code.
6. Publish packages.
7. Update packages in the required WasabiWallet project.

lontivero commented:
Lets discuss a bit about this before going too deep into it.

* This removes 10,000 LoC but adds 230,000 LoC and non-code artifacts like images. Having that extra amount of code doesn't feel okay because it is not free, i mean, it imposes additional costs to everybody from developers, CI, users (?) etc.

* How `git subtree` workflow intregrates with the existing Wasabi development environment is not obvious. How are subtrees treated by Visual Studio, JetBrains Rider, etc? Remember only linux devs use command line while the rest uses GUIs and a few ugly aliases that are not aware of subtrees.

* How does this play with the CI? is clonning a repo enough as before or it is necessary to fetch/add/clone subtrees too.

YohDeadfall commented:
Let's go in reverse order of questions.

> How does this play with the CI?

DeepSource is muted for any dependency we have, so it works as before, even better because there's no external code to analyze. Build times a bit increased because there are two more projects, but they are not so large to be noticeable.

> is clonning a repo enough as before or it is necessary to fetch/add/clone subtrees too.

Yes, the code is in our repository and is permanent. First two commits in this PR add ZXing code as is without any history from the original repository thankfully to `--squash` argument I pointed out in `README.md`

> How git subtree workflow intregrates with the existing Wasabi development environment is not obvious. How are subtrees treated by Visual Studio, JetBrains Rider, etc?

You don't need the command till you want to update that subtree. That's why it's preferable over submodules, i.e. there's nothing new to learn.

> Remember only linux devs use command line while the rest uses GUIs and a few ugly aliases that are not aware of subtrees.

It's a shell script shipped with git, so works as any other git command on any platform. Again, no one needs it till a subtree update.

> This removes 10,000 LoC but adds 230,000 LoC and non-code artifacts like images. Having that extra amount of code doesn't feel okay because it is not free, i mean, it imposes additional costs to everybody from developers, CI, users (?) etc.

Since it's just code copy pasted via the script, it's just code which you can edit as you want. Contributing back from a subtree could be problematic, but we don't plan to work this way.

The script downloads commits from an external repository and automatically makes two commits with standard messages. One of them points to the original commit in the source repository. Therefore, any developer can simply do `git subtree update` as in `README.md` and maybe resolve conflicts if they are (if we touch copy tree; otherwsie, nothing to do at all).

To reduce size of the ZXing I can remove tests and any assets there leaving just projects relevant to us. Next time we update ZXing any reviewer will see only differences between the source repository and our subtree.

To summarize, a subtree is just an automatically copied code from the external repository without any metadata, but with standard commit messages. If we need to reduce a subtree size we are free do to that.

YohDeadfall commented:
Taking into account the date of #7120, I assume `0.16.7.0` or `0.16.8.0` is used. There's [no big difference](https://github.com/micjahn/ZXing.Net/compare/v0.16.7.0...v0.16.8.0) between them. So, I can simply run `diff` between one of these versions and what we have to verify the assumption.

nopara73 reviewed (changes requested):
Very strong NACK. By introducing new systems to the Wasabi codebase, we leap into the abyss of the unknown, surrendering the security of our familiar existence to the whims of chance and entropy while creating the tower of Babel.

yahiheb reviewed (changes requested):
If this was the only way to have this scan QR codes feature I would vote for not having it at all for now.

NACK.

YohDeadfall commented:
@nopara73 and @yahiheb, please read my comments and proposals first. Then tell me what can be improved here. If it's about size, then there's an option to trim the code, but I want to hear that requested first.

Not approving without any constructive feedback isn't a positive way of making improvements especially on experimental stuff.

nopara73 reviewed (commented):
I withdraw my NACK here. I'd still prefer fork and NuGet approach, but if consensus can be achieved  on subtrees then I am not getting in the way.

YohDeadfall commented:
So, I added a very long description to the PR comparing different approaches which can be used. In addition to that I removed 100K of lines here just by `git rm` shell scripts by deleting tests and bindings we don't use anyway. Haven't touched the core library itself yet, but it should be pretty simple to. In the end, I guess, there will be about the same number of lines as were removed, but with an ability to run `git subtree update` to fetch changes from the original repository later.

YohDeadfall commented:
@yahiheb, @lontivero, the code size is almost identical to what is in `master`, but I decided to keep encoding too due to `WalletWasabi.Gma` used for QR code creation. I don't think that we need two libraries to deal with QR, so I would like to keep added files too. If it makes review harder, I can remove them now and return back when `Gma` time comes.

YohDeadfall commented:
Since ZXing supports RGB and YUV raw data, it's possible to avoid SkiaSharp in the future at all. That's why I left luminance sources for them.

molnard commented:
Great summary, thank you for putting that together. Subtree seems to be a nice function. 

I prefer NuGet as well - since we already put our leg there - we know what to expect - it works, easy to deploy. 

> Requires modification to the original code to use

That is fine, we are already having a reduced version of XZing. I believe small modifications in project files, etc. will be unavoidable. 

> Affects local and CI build time

Very strong benefits.

> Editing and committing.

We never want to do that. Stick with one working version. Only modify in case of a critical bug. 

> XZing code inside the repository

I do not see any benefit of that, but I see many drawbacks: analyzer/compiler warnings, unintentional change, IDE and my brain need to handle code instead of a library that I will never touch or review.

YohDeadfall commented:
> That is fine, we are already having a reduced version of XZing. I believe small modifications in project files, etc. will be unavoidable.

Yeah, in case of ZXing it was already so. I adopted both `csproj` files to our needs by removing non-interesting TFMs, signing and so on.

Moreover, I think we can nuke SkiaSharp from ZXing because the original version supports both RGB and YUV formats out of the box. Doing that will improve performance of camera input handling due to not allocating a Skia bitmap for QR code recognition.

> Very strong benefits.

Just run `dotnet build` on `master` and the current branch. The time is about the same in average, no noticeable difference - about 20 seconds in both cases. That's on a laptop with 16 GB of memory and AMD Ryzen 7 4700U CPU.

> I do not see any benefit of that, but I see many drawbacks: analyzer/compiler warnings, unintentional change, IDE and my brain need to handle code instead of a library that I will never touch or review.

DeepSource is muted for ZXing, so that's not an issue. IDE doesn't complain too just because each `git subtree` has its own `.editorconfig` as it's in a fork. From that point of view, it's even better than what we have in `master`.

molnard commented:
I've tested this. Cloning and checking out were without issues or any special commands. I have some questions:

- How can I check the subtree structure of the repo?
- How can I check which commit checked out in the subtree and what is the remote?

Having a Nuget package would make the developments and optimizations hard for you?

YohDeadfall commented:
> How can I check the subtree structure of the repo?

There's no sub-repository like it happens with `git submodule`. `git subtree` downloads code from the specified remote and produces commits in the current repository. Let's say, it's completely automated copy-pasta.

Updating a `git subtree` does almost the same thing.

> How can I check which commit checked out in the subtree and what is the remote?

There's no remote stored and no metadata, but you can get the original commit from the history: https://github.com/zkSNACKs/WalletWasabi/pull/10479/commits/ef099f91b4ad3acaf84457d6f21da547221a679b

> [Squashed 'Dependencies/ZXing.Net/' content from commit 3d870b70a](https://github.com/zkSNACKs/WalletWasabi/pull/10479/commits/ef099f91b4ad3acaf84457d6f21da547221a679b)
git-subtree-dir: Dependencies/ZXing.Net
git-subtree-split: 3d870b70af174a61ae45813d36ece7e811544d70

> Having a Nuget package would make the developments and optimizations hard for you?

I have to work in both repositories then, have rights to update packages (I would leave it to owners for security owners or at least grant the permission after a while), or tweak project files in `WalletWasabi` repository in case when I have to change a dependency in a fork (replacing `PackageReference` by `ProjectReference`) just to check that nothing important removed and the wallet compiles. So, one has to verify that I or anyone else won't push that tweak by an accident to a pull request.

lontivero commented:
Our policy about dependencies was born as a common sense reaction to the norm in software development to depend on an always growing number of packages from entrusted sources and the risks and consequences that such a high promiscuity practice entails.
       
All software is composed of already existing components and Wasabi is not the exception, we depend directly on NBitcoin, Avalonia, dotnet and indirectly we depend on hundreds or thousands of packages and tools. So, what is the limit or where do we draw a line? Should we also keep our own copy of openssl? Why is it okay to trust on Avalonia? 

The implicit idea is that we trust on some projects more that on others and we don't want to be worrying about those others so, we copied the code. It was always clear that we would never touch that code however we did it and we have done it many times simply because the code is there. Yes, some devs removed spaces, others changed namespaces, others removed methods that were not used, many had to review those useless changes and so on, all that even before we got addicted to the idiotic robots that complain about your lack of empty line before a comment, much before.

Code that will never change (and will not be change by us) is better to be reused in binary format. That should be also true even if we are the creators and maintainers, for example WabiSabi is now a nuget package and nobody is missing it in WalletWasabi, same thing happened with compact filters code that was originally in WalletWasabi but was moved to NBitcoin.

Same if valid here, if we put ZXing in a nuget package nobody is going to miss it, no additional code will need to be with us and we will enjoy the simplicity of having one and only one well-understood/known mechanism to deal with dependencies.

For me this is a NACK.

YohDeadfall commented:
For me it's fine to replace such things with fine NuGet packages which we don't own. Only one reason why I did that work was a security concern issued previously by other members.

cc @nopara73

YohDeadfall commented:
Misunderstood you, @lontivero. So, if I'm correct you are suggesting to make a fork and publish a package built from it and prefixed by `WalletWasabi`, right?

nopara73 commented:
Lucas meant NuGet setup that we own, not a NuGet setup like Avalonia or NBitcoin, so no security concerns there. Also since you've already done the useless code removal we may just come up with a name for this NuGet like QRoissant or QRackers to keep up with the zkSNACKs' restaurant naming schemes (Wasabi, Dumplings)

nopara73 commented:
QRackers sounds really cool. Now I want it!!!

YohDeadfall commented:
There's one more thing I have in my mind which can work well and doesn't require forks. It's `Directory.Build.props` (or whatever we call it) containing a centralized list of packages versions and a code owners file setting up reviewers for NuGet lock files. Then any version change must be approved at least by one of the members listed there.

For simplicity there can be two files for versions, one for trusted and another one for less trusted packages. It can be changed since the core idea is protection of lock files by code owners.

Benefits and downsides as for any NuGet package. No forks, nothing to maintain, no security issues.

Is it sound?

YohDeadfall commented:
The proposed solution with code owners also helps to secure package addition. If it happens, then the core team must review it first.

molnard commented:
I believe everyone thought this over, twice and the positions are clear. Nuget is the way to go. 

Here is the example repo, we should do it exactly the same way: https://github.com/zkSNACKs/NWabiSabi

Created the new repo here, you are Admin there @YohDeadfall https://github.com/zkSNACKs/NQRackers

YohDeadfall commented:
@molnard, my last option about NuGet too, but without forking.

> Created the new repo here

Won't it be an issue that we take someone's work, drop some code and publish it under a new name surely with the original license? Asking it from a legal point of view.

molnard commented:
> code owners file setting up reviewers for NuGet lock files

I would not add more complexity. The "system" works fine. Let's switch to executive mode with this task. Our users wanna use QR reader on Linux. BTW Mac users are taking the lead and they still cannot use this extremely useful feature.

YohDeadfall commented:
> BTW Mac users are taking the lead and they still cannot use this extremely useful feature.

Well, that's a wrong library then. ZXing just recognizes QR codes from supplied images, but to get images on Mac I have to work on FlashCap.

molnard commented:
> Won't it be an issue that we take someone's work, drop some code and publish it under a new name surely with the original license? Asking it from a legal point of view.

Good point I do not know. Please do research. As far as I remember it was enough to copy their license next to the source code.

molnard commented:
> Well, that's a wrong library then. ZXing just recognizes QR codes from supplied images, but to get images on Mac I have to work on FlashCap.

It is OK. The vision is to be cross-platform with every feature. We are going step by step.

molnard commented:
> drop some code and publish it under a new name

The name can be changed. However, I think naming it similarly then the original is worse and just causes confusion to whoever looking for the original implementation on Nuget. I would not do that personally.

molnard commented:
@YohDeadfall you are a God (Admin) in this repository, power comes with responsibility. Use your force wisely. :)

You can rename, change anything, etc. I believe at some point you will need Lucas for the token but other than that you can do everything. 
https://github.com/zkSNACKs/NQRackers

YohDeadfall commented:
> Good point I do not know. Please do research. As far as I remember it was enough to copy their license next to the source code.

Apache License 2.0 allows to make modifications, but with keeping the original license. The problem is that it's empty and has no real owner specified in it. Not a lawyer to understand consequences for us.

> The name can be changed. However, I think naming it similarly then the original is worse and just causes confusion to whoever looking for the original implementation on Nuget. I would not do that personally.

It's a common practice for companies to prefix someones work with their own product or company name which tells the user to whom it belongs and not to expect updates in that library since it's private.

Giving a new name without telling that it's a part of Wasabi would attract users to depend on it and later to contribute. The exact thing we would like to avoid.

It's possible to make it completely private by using MyGet stream, so the package won't be listed on NuGet and found by users at all. What is needed on our side then is making a `NuGet.config` with addition of our private stream on MyGet.

YohDeadfall commented:
MyGet is paid, so doesn't look attractive. Perhaps, Shay used a basic plan for Npgsql, but maybe it was completely free before.

molnard commented:
> Apache License 2.0 allows to make modifications, but with keeping the original license.

Perfectly enough. 

> It's a common practice for companies to prefix someones work

Not in our use case. We do not want to make a better XZing or add more functionality for others. If the name does not contain the original XZing name, nobody will find it and try. It is enough to write down this in the readme. Your concerns are perfectly valid, we must take them into account => naming should not look like XZing. 

> MyGet stream

No reason. Well-written readme inside our repo is good enough.

YohDeadfall commented:
> Perfectly enough.

Except that there's a template, so it can be an invalid license and a potential issue for the company. If the company owners are fine with taking that, then I will start on the new repo.

@MaxHillebrand @nopara73

YohDeadfall commented:
Checked the license once more. The project add copyright to each file where all information is provided, so there's a valid licence and we can copy it.
